plugins/inputs/modbus/modbus.go

```diff
@@ -173,552 +167,253 @@ func (m *Modbus) Init() error {
 		return fmt.Errorf("retries cannot be negative")
 	}
 
-	err := m.InitRegister(m.DiscreteInputs, cDiscreteInputs)
-	if err != nil {
-		return err
+	// Check and process the configuration
+	if err := m.ConfigurationOriginal.Check(); err != nil {
+		return fmt.Errorf("original configuraton invalid: %v", err)
 	}
 
-	err = m.InitRegister(m.Coils, cCoils)
+	r, err := m.ConfigurationOriginal.Process()
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot process original configuraton: %v", err)
 	}
+	m.requests = r
 
-	err = m.InitRegister(m.HoldingRegisters, cHoldingRegisters)
-	if err != nil {
-		return err
-	}
-
-	err = m.InitRegister(m.InputRegisters, cInputRegisters)
-	if err != nil {
-		return err
+	// Setup client
+	if err := m.initClient(); err != nil {
+		return fmt.Errorf("initializing client failed: %v", err)
 	}
 
 	return nil
 }
 
-func (m *Modbus) InitRegister(fields []fieldContainer, name string) error {
-	if len(fields) == 0 {
-		return nil
-	}
-
-	err := validateFieldContainers(fields, name)
-	if err != nil {
-		return err
+// Gather implements the telegraf plugin interface method for data accumulation
+func (m *Modbus) Gather(acc telegraf.Accumulator) error {
+	if !m.isConnected {
+		if err := m.connect(); err != nil {
+			return err
+		}
 	}
 
-	addrs := []uint16{}
-	for _, field := range fields {
-		addrs = append(addrs, field.Address...)
+	timestamp := time.Now()
+	for retry := 0; retry <= m.Retries; retry++ {
+		timestamp = time.Now()
+		if err := m.gatherFields(); err != nil {
+			if mberr, ok := err.(*mb.Error); ok && mberr.ExceptionCode == mb.ExceptionCodeServerDeviceBusy && retry < m.Retries {
+				m.Log.Infof("Device busy! Retrying %d more time(s)...", m.Retries-retry)
+				time.Sleep(time.Duration(m.RetriesWaitTime))
```

Comment:
probably inherited code, but we might want to consider putting non-responsive devices into a list and processing that list over until it's empty, rather than sleeping mid-list.

Reply:
Well the plugin can handle only one device and if that device is busy we cannot do anything in the meantime... I thought about putting off the sleep but it adds complexity for no gain IMO.